Number_types/include/CGAL/cpp_float.h

```diff
@@ -0,0 +1,496 @@
+
+    uint64_t m;
+    uint64_t dexp = u.s.exp;
+    CGAL_assertion_msg(dexp != 2047, "Creating an cpp_float from infinity or NaN.");
+    if (dexp == 0) {
+      if (d == 0) { exp=0; return; }
+      else { // denormal number
+        m = u.s.man;
+        ++dexp;
+      }
+    } else {
+      m = (1LL << 52) |  u.s.man;
+    }
+
+
+    int idexp = (int)dexp;
+    idexp -= 1023;
+
+    // std::cout << "m    = "  << m << std::endl;
+    // std::cout << "idexp = " << idexp << std::endl;
+
+    int shifted = internal::low_bit(m);
+
+    m >>= shifted;
+
+    int nbits = internal::high_bit(m);
+    // std::cout << "nbits = " << nbits << std::endl;
+
+    exp = idexp - nbits;
+    if(u.s.sig){
+      m = -m;
+    }
+    // std::cout << "m = " << m << " * 2^" << exp  << std::endl;
+    // fmt(m);
+    man = m;
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const cpp_float& m)
+  {
+    return os << m.man << " * 2 ^ " << m.exp << " ( " << m.to_double() << ") ";
+  }
+
+  friend cpp_float operator-(cpp_float const&x)
+  {
+#ifdef CGAL_CPPF
+    return cpp_float(-x.man,x.exp, -x.rat);
+#else
+    return cpp_float(-x.man,x.exp);
+#endif
+  }
+
+  cpp_float& operator*=(const cpp_float& other)
+  {
+#ifdef CGAL_CPPF
+    rat *= other.rat;
+#endif
+    man *= other.man;
+    exp += other.exp;
+    return *this;
+  }
+
+
+  friend
+  cpp_float operator*(const cpp_float& a, const cpp_float&b){
+    return cpp_float(a.man*b.man, a.exp+b.exp);
+  }
+
+
+  cpp_float operator+=(const cpp_float& other)
+  {
+#ifdef CGAL_CPPF
+    rat += other.rat;
+#endif
+    int shift = exp - other.exp;
+    if(shift > 0){
+      man <<= shift;
+      man += other.man;
+      exp = other.exp;
+    }else if(shift < 0){
+      man += (other.man << -shift);
+    }else{
+      man += other.man;
+    }
+    return *this;
+  }
+
+
+  friend
+  cpp_float operator+(const cpp_float& a, const cpp_float&b){
+    int shift = a.exp - b.exp;
+    if(shift > 0){
+      return cpp_float((a.man << shift) + b.man, b.exp);
+    }else if(shift < 0){
+      return cpp_float(a.man + (b.man << -shift), a.exp);
+    }
+    return cpp_float(a.man + b.man, a.exp);
+  }
+
+
+
+  cpp_float operator-=(const cpp_float& other)
+  {
+
+#ifdef CGAL_CPPF
+    rat -= other.rat;
+#endif
+    int shift = exp - other.exp;
+    if(shift > 0){
+      man <<= shift;
+      man -= other.man;
+      exp = other.exp;
+    }else if(shift < 0){
+      man -= (other.man << -shift);
+    }else{
+      man -= other.man;
+    }
+    return *this;
+  }
+
+  friend
+  cpp_float operator-(const cpp_float& a, const cpp_float&b){
+
+    int shift = a.exp - b.exp;
+    if(shift > 0){
+      return cpp_float((a.man << shift) - b.man, b.exp);
+    }else if(shift < 0){
+      return cpp_float(a.man - (b.man << -shift), a.exp);
+    }
+    return cpp_float(a.man - b.man, a.exp);
+  }
+
+  bool positive() const
+  {
+    return is_positive(man);
+  }
+
+
+  friend bool operator<(const cpp_float& a, const cpp_float& b)
+  {
```

Comment:
It makes sense to first compare  just the sign.   And then the sign of the exponent. And then only something expensive.